### PR TITLE
how to register a rely function Issue-478

### LINKED ISSUE #478: how do i get each agent message? instead of agent last message?
how do i get each agent message? instead of agent last message?

### LINKED ISSUE #559: Examples of how to register a reply function
@xmi1an you can hook in function that is called whenever an agent receives a message. 

Assuming you have two agents `user_proxy` and `assistant`, you can [register a function](https://microsoft.github.io/autogen/docs/reference/agentchat/conversable_agent#conversableagent-objects) with these agents that gets called with some parameters. 

```python

def print_messages(recipient, messages, sender, config): 
    if "callback" in config and  config["callback"] is not None:
        callback = config["callback"]
        callback(sender, recipient, messages[-1])
    print(f"Messages sent to: {recipient.name} | num messages: {len(messages)}")
    return False, None  # required to ensure the agent communication flow continues

user_proxy.register_reply(
    [autogen.Agent, None],
    reply_func=print_messages, 
    config={"callback": None},
)

assistant.register_reply(
    [autogen.Agent, None],
    reply_func=print_messages, 
    config={"callback": None},
) 
``` 

In the above, we register a `print_messages` message that is called each time the agent receive a message.

_Originally posted by @victordibia in https://github.com/microsoft/autogen/issues/478#issuecomment-1785922097_

It's a good example to add to FAQ.

### PR DESCRIPTION
**Pull request by olaoluwasalami**
@sonichi @victordibia @thinkall 

## Why are these changes needed?

FAQ: How to register a reply function

## Related issue number

Closes #559 

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by olaoluwasalami** [comment]
> @microsoft-github-policy-service agree

@microsoft-github-policy-service agree

**Review by thinkall — COMMENTED** [review]
Thank you very much, @olaoluwasalami ! 

It's worth mentioning more use cases for this, for example, adding a `send_message_to_ui` function that sends the received message to the UI or a remote endpoint.

@sonichi , this could be a handy way to address some issues raised in the community. We should emphasize this function. WDYT?

File: website/docs/FAQ.md [hunk 1]
```diff
@@ -127,3 +127,34 @@ Workaround:
...
 
 Explanation: Per [this gist](https://gist.github.com/defulmere/8b9695e415a44271061cc8e272f3c300?permalink_comment_id=4711478#gistcomment-4711478), linked from the official [chromadb docs](https://docs.trychroma.com/troubleshooting#sqlite), adding this folder triggers chromadb to use pysqlite3 instead of the default.
+
+## How to register a reply function
+
+(from [issue #478](https://github.com/microsoft/autogen/issues/478))
+
+You can hook in a function that is called whenever an agent receives a message
+
+Assuming you have two agents user_proxy and assistant, you can register a function with these agents that gets called with some parameters.
```
Comment (afourney) [inline website/docs/FAQ.md:137]: "you can register a function with these agents that gets called with some parameters" is a bit too generic on its own. Perhaps link to the documentation: https://microsoft.github.io/autogen/docs/reference/agentchat/conversable_agent#register_reply

Comment (olaoluwasalami) [reply website/docs/FAQ.md:137]: Thanks @afourney

Comment (sonichi) [reply website/docs/FAQ.md:137]: I agree with adding a link to the documentation.

File: website/docs/FAQ.md [hunk 2]
```diff
@@ -127,3 +127,34 @@ Workaround:
 2. `mkdir /home/vscode/.local/lib/python3.10/site-packages/google/colab`
 
 Explanation: Per [this gist](https://gist.github.com/defulmere/8b9695e415a44271061cc8e272f3c300?permalink_comment_id=4711478#gistcomment-4711478), linked from the official [chromadb docs](https://docs.trychroma.com/troubleshooting#sqlite), adding this folder triggers chromadb to use pysqlite3 instead of the default.
+
+## How to register a reply function
+
+(from [issue #478](https://github.com/microsoft/autogen/issues/478))
+
+You can hook in a function that is called whenever an agent receives a message
```
Comment (sonichi) [inline website/docs/FAQ.md:135]: The registered reply function is only called when `generate_reply` is called. If an agent receive a message but a reply is not requested, it won't be called.

File: website/docs/FAQ.md [hunk 3]
```diff
@@ -127,3 +127,34 @@ Workaround:
...
+)
+
+assistant.register_reply(
+    [autogen.Agent, None],
+    reply_func=print_messages,
+    config={"callback": None},
+)
+
+```
+In the above, we register a print_messages message that is called each time the agent receive a message.
```
Comment (sonichi) [inline website/docs/FAQ.md:160]: This is not accurate. See my comment above.

**Review by sonichi — COMMENTED** [review]
I agree with @thinkall about the importance. It's worth a more detailed documentation or a blog post. It's OK to have this section here in FAQ before that.

**Comment by olaoluwasalami** [comment]
@sonichi Please review.

File: website/docs/FAQ.md [hunk 4]
```diff
@@ -132,6 +132,12 @@ Workaround:
 
 Explanation: Per [this gist](https://gist.github.com/defulmere/8b9695e415a44271061cc8e272f3c300?permalink_comment_id=4711478#gistcomment-4711478), linked from the official [chromadb docs](https://docs.trychroma.com/troubleshooting#sqlite), adding this folder triggers chromadb to use pysqlite3 instead of the default.
 
+## How to register a reply function
+
+(from [issue #478](https://github.com/microsoft/autogen/issues/478))
```
Comment (sonichi) [inline website/docs/FAQ.md:137]: What about copying the example from #478 to here?
@victordibia @gagb what do you think?

Comment (JieyuZ2) [reply website/docs/FAQ.md:137]: I think directly show the example would be good for readers

Comment (olaoluwasalami) [reply website/docs/FAQ.md:137]: Example added

File: website/docs/FAQ.md [hunk 5]
```diff
@@ -132,6 +132,37 @@ Workaround:
...
+    config={"callback": None},
+)
+
+assistant.register_reply(
+    [autogen.Agent, None],
+    reply_func=print_messages,
+    config={"callback": None},
+)
+```
+In the above, we register a print_messages message that is called each time the agent receive a message.
```
Comment (JieyuZ2) [inline website/docs/FAQ.md:164]: print_messages message -> `print_messages` function

Comment (olaoluwasalami) [reply website/docs/FAQ.md]: fixed @JieyuZ2

**Review by qingyun-wu — APPROVED** [review]
LGTM! Thanks!

File: website/docs/FAQ.md [hunk 6]
```diff
@@ -132,6 +132,37 @@ Workaround:
...
 
+## How to register a reply function
+
+(from [issue #478](https://github.com/microsoft/autogen/issues/478))
+
+See here https://microsoft.github.io/autogen/docs/reference/agentchat/conversable_agent/#register_reply
+
+For example,
+
+ Assuming you have two agents user_proxy and assistant, you can register a function with these agents that gets called with some parameters. (The registered reply function is only called when `generate_reply` is called)
```
Comment (sonichi) [inline website/docs/FAQ.md:143]:
```suggestion
For example, you can register a reply function that gets called when `generate_reply` is called for an agent.
```

File: website/docs/FAQ.md [hunk 7]
```diff
@@ -132,6 +132,37 @@ Workaround:
...
+    config={"callback": None},
+)
+
+assistant.register_reply(
+    [autogen.Agent, None],
+    reply_func=print_messages,
+    config={"callback": None},
+)
+```
+In the above, we register a `print_messages` function that is called each time the agent receive a message.
```
Comment (sonichi) [inline website/docs/FAQ.md:164]:
```suggestion
In the above, we register a `print_messages` function that is called each time the agent's `generate_reply` is triggered after receiving a message.
```

Comment (olaoluwasalami) [reply website/docs/FAQ.md]: changes made

**Review by sonichi — APPROVED** [review]

**Comment by olaoluwasalami** [comment]
@sonichi  auto-merge disabled

**Pull request merged**